<a href="https://colab.research.google.com/github/KristineYW/DS-Unit-2-Linear-Models/blob/master/Kristine_LS_DS_212_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 2*

---

# Regression 2

## Assignment

You'll continue to **predict how much it costs to rent an apartment in NYC,** using the dataset from renthop.com.

- [ ] Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
- [ ] Engineer at least two new features. (See below for explanation & ideas.)
- [ ] Fit a linear regression model with at least two features.
- [ ] Get the model's coefficients and intercept.
- [ ] Get regression metrics RMSE, MAE, and $R^2$, for both the train and test data.
- [ ] What's the best test MAE you can get? Share your score and features used with your cohort on Slack!
- [ ] As always, commit your notebook to your fork of the GitHub repo.


#### [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)

> "Some machine learning projects succeed and some fail. What makes the difference? Easily the most important factor is the features used." — Pedro Domingos, ["A Few Useful Things to Know about Machine Learning"](https://homes.cs.washington.edu/~pedrod/papers/cacm12.pdf)

> "Coming up with features is difficult, time-consuming, requires expert knowledge. 'Applied machine learning' is basically feature engineering." — Andrew Ng, [Machine Learning and AI via Brain simulations](https://forum.stanford.edu/events/2011/2011slides/plenary/2011plenaryNg.pdf) 

> Feature engineering is the process of using domain knowledge of the data to create features that make machine learning algorithms work. 

#### Feature Ideas
- Does the apartment have a description?
- How long is the description?
- How many total perks does each apartment have?
- Are cats _or_ dogs allowed?
- Are cats _and_ dogs allowed?
- Total number of rooms (beds + baths)
- Ratio of beds to baths
- What's the neighborhood, based on address or latitude & longitude?

## Stretch Goals
- [ ] If you want more math, skim [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf),  Chapter 3.1, Simple Linear Regression, & Chapter 3.2, Multiple Linear Regression
- [ ] If you want more introduction, watch [Brandon Foltz, Statistics 101: Simple Linear Regression](https://www.youtube.com/watch?v=ZkjP5RJLQF4)
(20 minutes, over 1 million views)
- [ ] Add your own stretch goal(s) !

In [0]:
%%capture
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'
    !pip install category_encoders==2.*

# If you're working locally:
else:
    DATA_PATH = '../data/'
    
# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [0]:
import numpy as np
import pandas as pd

# Read New York City apartment rental listing data
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv')
assert df.shape == (49352, 34)

# Remove the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= np.percentile(df['price'], 0.5)) & 
        (df['price'] <= np.percentile(df['price'], 99.5)) & 
        (df['latitude'] >= np.percentile(df['latitude'], 0.05)) & 
        (df['latitude'] < np.percentile(df['latitude'], 99.95)) &
        (df['longitude'] >= np.percentile(df['longitude'], 0.05)) & 
        (df['longitude'] <= np.percentile(df['longitude'], 99.95))]

 #### Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.


In [313]:
# Look at the data 

df.head()

bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0
2        1.0         1  ...                 0                    0
3        1.0         1  ...                 0                    0
4        1.0         4  ...                 0                    0

[5 rows x 34 columns]

In [314]:
# Make sure that "created" data is in datetime format

df.dtypes

bathrooms               float64
bedrooms                  int64
created                  object
description              object
display_address          object
latitude                float64
longitude               float64
price                     int64
street_address           object
interest_level           object
elevator                  int64
cats_allowed              int64
hardwood_floors           int64
dogs_allowed              int64
doorman                   int64
dishwasher                int64
no_fee                    int64
laundry_in_building       int64
fitness_center            int64
pre-war                   int64
laundry_in_unit           int64
roof_deck                 int64
outdoor_space             int64
dining_room               int64
high_speed_internet       int64
balcony                   int64
swimming_pool             int64
new_construction          int64
terrace                   int64
exclusive                 int64
loft                      int64
garden_p

In [315]:
# Convert column to datetime format

df['created'] = df['created'].astype('datetime64')

df.head()

bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0
2        1.0         1  ...                 0                    0
3        1.0         1  ...                 0                    0
4        1.0         4  ...                 0                    0

[5 rows x 34 columns]

In [316]:
# Check to make sure conversion worked for datetime

df.dtypes

bathrooms                      float64
bedrooms                         int64
created                 datetime64[ns]
description                     object
display_address                 object
latitude                       float64
longitude                      float64
price                            int64
street_address                  object
interest_level                  object
elevator                         int64
cats_allowed                     int64
hardwood_floors                  int64
dogs_allowed                     int64
doorman                          int64
dishwasher                       int64
no_fee                           int64
laundry_in_building              int64
fitness_center                   int64
pre-war                          int64
laundry_in_unit                  int64
roof_deck                        int64
outdoor_space                    int64
dining_room                      int64
high_speed_internet              int64
balcony                  

In [317]:
# Pull the months out into their own columns 

df['month'] = df['created'].dt.month

df.head()

bathrooms  bedrooms  ... common_outdoor_space month
0        1.5         3  ...                    0     6
1        1.0         2  ...                    0     6
2        1.0         1  ...                    0     4
3        1.0         1  ...                    0     4
4        1.0         4  ...                    0     4

[5 rows x 35 columns]

In [0]:
# Separate months into its own column

df['month'] = df['month'].astype(str)

# Create datasets for test and train data

train = df[(df['month'] == '5') | (df['month'] =='4')]
test = df[df['month'] == '6']

In [319]:
# Check how many observations there are 

train.shape,test.shape

((31844, 35), (16973, 35))

 #### Engineer at least two new features.



In [320]:
# Feature 1: Length of description. Create a new column with the length of the description

# Create new column with length of the description column 

df['len_desc'] = df['description'].str.len()

df.head()

bathrooms  bedrooms  ... month len_desc
0        1.5         3  ...     6    588.0
1        1.0         2  ...     6      8.0
2        1.0         1  ...     4    691.0
3        1.0         1  ...     4    492.0
4        1.0         4  ...     4    479.0

[5 rows x 36 columns]

In [321]:
# Feature 2: Total number of rooms, bathroom + bedroom

# Create new variable with sum value 
sum_rooms = df['bathrooms'] + df['bedrooms']

# Assign variable to new column
df['total_rooms'] = sum_rooms

# Check rows for new column
df.head()

bathrooms  bedrooms             created  ... month len_desc  total_rooms
0        1.5         3 2016-06-24 07:54:24  ...     6    588.0          4.5
1        1.0         2 2016-06-12 12:19:27  ...     6      8.0          3.0
2        1.0         1 2016-04-17 03:26:41  ...     4    691.0          2.0
3        1.0         1 2016-04-18 02:22:02  ...     4    492.0          2.0
4        1.0         4 2016-04-28 01:32:41  ...     4    479.0          5.0

[5 rows x 37 columns]

 #### Fit a linear regression model with at least two features.


In [0]:
# Import estimator from Scikit-Learn

from sklearn.linear_model import LinearRegression
import plotly.express as px


In [0]:
# Instantiate the class

model = LinearRegression()

In [324]:
# Use sklearn to fit a multiple regression with the two new features

features = ['bedrooms','bathrooms']
print(f'Linear Regression, dependent on: {features}')

X_train = train[features]
X_test = test[features]

Linear Regression, dependent on: ['bedrooms', 'bathrooms']


In [0]:
# Train the y target vector

target = 'price'
y_train = train[target]
y_test = test[target]

In [326]:
# Fit the model

model.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [0]:
# Apply the model to new data

y_pred = model.predict(X_train)

In [328]:
# Find the equation for the plane 

beta0 = model.intercept_
beta1, beta2 = model.coef_
print(f'y = {beta0} + {beta1}x1 + {beta2}x2')

y = 485.71869002322273 + 389.32489590255824x1 + 2072.6101163851895x2


In [329]:
# Make the betas/model correlations easier to understand

print('Intercept', model.intercept_)
coefficients = pd.Series(model.coef_, features)
print(coefficients.to_string())

Intercept 485.71869002322273
bedrooms      389.324896
bathrooms    2072.610116


 #### Get the model's coefficients and intercept.


In [330]:
# Use methods to find intercept

model.intercept_

485.71869002322273

In [331]:
# Use methods to find coefficient

model.coef_

array([ 389.3248959 , 2072.61011639])

 #### Get regression metrics RMSE, MAE, and  𝑅2 , for both the train and test data.


In [0]:
# Import libraries

import numpy as np
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score


In [333]:
# Calculate the MAE for the training data

mae = mean_absolute_error(y_train,y_pred)

mae

818.5310213271714

In [334]:
# Calculate the MAE for the testing data

y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test,y_pred)

mae

825.8987822403527

In [335]:
# Calculate the RMSE for the training data

import math

y_pred = model.predict(X_train)
mse = mean_squared_error(y_train,y_pred)
rmse = math.sqrt(mse)

rmse

1232.0225917223486

In [336]:
# Calculate the RMSE for the test data


y_pred = model.predict(X_test)
mse = mean_squared_error(y_test,y_pred)
rmse = math.sqrt(mse)

rmse

1219.719357233823

In [337]:
# Calculate the R-squared for the training data

y_pred = model.predict(X_train)
r2_score(y_train,y_pred)

0.5111543084316605

In [338]:
# Calculate the R-squared for the test data

y_pred = model.predict(X_test)
r2_score(y_test,y_pred)

0.5213303957090346

#### What's the best test MAE you can get? Share your score and features used with your cohort on Slack!

After running all the cells again with different sets of features, the best MAE I was able to get was for the set of bathrooms, bedrooms, and the test MAE was 825.9.

#### As always, commit your notebook to your fork of the GitHub repo.